public/app/features/expressions/components/Threshold.tsx

```diff
@@ -1,70 +1,85 @@
-import { ButtonSelect, InlineField, InlineFieldRow, Input, Select, useStyles2 } from '@grafana/ui';
+import { Stack } from '@grafana/experimental';
+import { ButtonSelect, InlineField, InlineFieldRow, InlineSwitch, Input, Select, useStyles2 } from '@grafana/ui';
+import { config } from 'app/core/config';
 import { EvalFunction } from 'app/features/alerting/state/alertDef';
 
 import { ClassicCondition, ExpressionQuery, thresholdFunctions } from '../types';
 
+import {
+  isInvalid,
+  thresholdReducer,
+  updateHysteresisChecked,
+  updateRefId,
+  updateThresholdParams,
+  updateThresholdType,
+  updateUnloadParams,
+} from './thresholdReducer';
+
 interface Props {
   labelWidth: number | 'auto';
   refIds: Array<SelectableValue<string>>;
   query: ExpressionQuery;
   onChange: (query: ExpressionQuery) => void;
+  onError?: (error: string | undefined) => void;
 }
 
 const defaultThresholdFunction = EvalFunction.IsAbove;
 
-export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
+const defaultEvaluator: ClassicCondition = {
+  type: 'query',
+  evaluator: {
+    type: defaultThresholdFunction,
+    params: [0, 0],
+  },
+  query: {
+    params: [],
+  },
+  reducer: {
+    params: [],
+    type: 'last',
+  },
+};
+
+export const Threshold = ({ labelWidth, onChange, refIds, query, onError }: Props) => {
   const styles = useStyles2(getStyles);
 
-  const defaultEvaluator: ClassicCondition = {
-    type: 'query',
-    evaluator: {
-      type: defaultThresholdFunction,
-      params: [0, 0],
-    },
-    query: {
-      params: [],
-    },
-    reducer: {
-      params: [],
-      type: 'last',
-    },
-  };
+  const initialExpression = { ...query, conditions: query.conditions?.length ? query.conditions : [defaultEvaluator] };
 
-  const conditions = query.conditions?.length ? query.conditions : [defaultEvaluator];
-  const condition = conditions[0];
+  // this queryState is the source of truth for the threshold component.
+  // All the changes are made to this object through the dispatch function with the thresholdReducer.
+  const [queryState, dispatch] = useReducer(thresholdReducer, initialExpression);
+  const conditionInState = queryState.conditions[0];
 
-  const thresholdFunction = thresholdFunctions.find((fn) => fn.value === conditions[0].evaluator?.type);
+  const thresholdFunction = thresholdFunctions.find((fn) => fn.value === queryState.conditions[0].evaluator?.type);
 
   const onRefIdChange = (value: SelectableValue<string>) => {
-    onChange({ ...query, expression: value.value });
+    dispatch(updateRefId(value.value));
   };
 
-  const onEvalFunctionChange = (value: SelectableValue<EvalFunction>) => {
-    const type = value.value ?? defaultThresholdFunction;
+  // any change in the queryState will trigger the onChange function.
+  useEffect(() => {
+    queryState && onChange(queryState);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I was unable to remove this one. It appears that `onChange` is being updated each time it's called. While I managed to address it when used in our alerting component by adding a `useCallback` in the parent, I couldn't find a way to implement useCallback when used in the `ExpressionQueryEditor` because `onChange` comes from the `ExpressionDataSource` code. This is the reason for reverting the last commit.